### PR TITLE
140040 - BESS builder service get permit key create key file

### PR DESCRIPTION
**Pull request by MastekRohitChettri**
This PR consist changes of 
[PBI : 140040](https://dev.azure.com/ukhydro/Exchange%20Set%20Service/_workitems/edit/140040)
[PBI : 156675](https://dev.azure.com/ukhydro/Exchange%20Set%20Service/_workitems/edit/156675)

- Added code changes
- Added unit tests
- Added functional test cases
- Added required variable for appsetting
- Change AD Authorization for PKS from app registration to managed identity.

### CODE REVIEW

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Fulfilment.UnitTests/Helpers/FileSystemHelperTests.cs
Comment (Nadeem14456) [inline UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Fulfilment.UnitTests/Helpers/FileSystemHelperTests.cs:1]: Follow When_Then naming convention for testcases

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/UKHO.PeriodicOutputService.Common.csproj
Comment (Nadeem14456) [inline UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/UKHO.PeriodicOutputService.Common.csproj:1]: Note: suppressions are restored from main branch

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Services/PksService.cs
Comment (Nadeem14456) [inline UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Services/PksService.cs:1]: remove blank space at line 19

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/PermitDecryption/S63Crypt.cs [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+            for (int i = 0; i < data.Length; i++)
+            {
+                b = data[i] >> 4;
+                c[i * 2] = (char)(55 + b + (((b - 10) >> 31) & -7));
+                b = data[i] & 0xF;
+                c[i * 2 + 1] = (char)(55 + b + (((b - 10) >> 31) & -7));
+            }
+            return new string(c);
+        }
+
```
Comment (Nadeem14456) [inline UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/PermitDecryption/S63Crypt.cs:93]: remove blank space

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/PermitDecryption/BlowfishFactory.cs [hunk 2]
```diff
@@ -0,0 +1,53 @@
+﻿using System.Diagnostics.CodeAnalysis;
+using System.Runtime.Caching;
+
+namespace UKHO.PeriodicOutputService.Common.PermitDecryption
+{
+
```
Comment (Nadeem14456) [inline UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/PermitDecryption/BlowfishFactory.cs:6]: remove blank space

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/PermitDecryption/BlowfishFactory.cs [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    [ExcludeFromCodeCoverage]
+    public class BlowfishFactory : IBlowfishFactory
+    {
```
Comment (Nadeem14456) [inline UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/PermitDecryption/BlowfishFactory.cs:9]:
```suggestion
    {
    
```

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/PermitDecryption/BlowfishAlgorithm.cs
Comment (Nadeem14456) [inline UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/PermitDecryption/BlowfishAlgorithm.cs:1]: inconsistent formatting

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Models/PermitKey.cs [hunk 4]
```diff
@@ -0,0 +1,12 @@
+﻿using System.Diagnostics.CodeAnalysis;
+
+namespace UKHO.PeriodicOutputService.Common.Models
+{
+    [ExcludeFromCodeCoverage]
+    public class PermitKey
+    {
+        public string ActiveKey { get; set; }
+        public string NextKey { get; set; }
+
```
Comment (Nadeem14456) [inline UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Models/PermitKey.cs:10]: remove blank space

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Models/Bess/ProductKeyServiceResponse.cs [hunk 5]
```diff
@@ -1,17 +1,21 @@
-﻿using Newtonsoft.Json;
+﻿using System.Xml.Serialization;
+using Newtonsoft.Json;
 using Newtonsoft.Json.Serialization;
```
Comment (Nadeem14456) [inline UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Models/Bess/ProductKeyServiceResponse.cs:3]: unused using
```suggestion
using Newtonsoft.Json;
```

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Models/Bess/PKSXml.cs [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+        public string noNamespaceSchemaLocation = ".\\UKHO.PeriodicOutputService\\UKHO.PeriodicOutputService.Common\\XmlSchema\\CellPermitExchange.xsd";
+    }
+
+    [ExcludeFromCodeCoverage]
+    [XmlRoot(ElementName = "cellkeys")]
+    public class cellkeys
+    {
+        [XmlElement(ElementName = "cell")]
+        public List<ProductKeyServiceResponse> response { get; set; }
+    }
```
Comment (Nadeem14456) [inline UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Models/Bess/PKSXml.cs:26]: move this to a separate class.

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Models/Bess/PKSXml.cs
Comment (Nadeem14456) [inline UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Models/Bess/PKSXml.cs:1]: Use Pascal case for class & fields

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Logging/EventIds.cs [hunk 7]
```diff
@@ -773,7 +773,12 @@ public enum EventIds
         /// <summary>
         /// 809153 - Post product data to Product Key Service failed.
         /// </summary>
-        PostProductKeyDataToPksFailed = 809153
+        PostProductKeyDataToPksFailed = 809153,
+
+        /// <summary>
+        /// 809154 - Permit Decryption Exception
```
Comment (Nadeem14456) [inline UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Logging/EventIds.cs:779]:
```suggestion
        /// 809154 - Exception occurred while Permit Decryption
```

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/IFileSystemHelper.cs
Comment (Nadeem14456) [inline UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/IFileSystemHelper.cs:1]: please remove blank space at line 31 & add at line 38

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/AuthTokenProvider.cs [hunk 8]
```diff
@@ -29,7 +29,7 @@ public class AuthTokenProvider : IAuthEssTokenProvider, IAuthFssTokenProvider, I
             _essManagedIdentityConfiguration = essManagedIdentityConfiguration;
             _cache = cache;
             _logger = logger;
-            _pksApiConfiguration = pksApiConfiguration.Value ?? throw new ArgumentNullException(nameof(pksApiConfiguration)); ;
+            this._pksApiConfiguration = pksApiConfiguration.Value ?? throw new ArgumentNullException(nameof(pksApiConfiguration)); ;
```
Comment (Nadeem14456) [inline UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/AuthTokenProvider.cs:32]:
```suggestion
            _pksApiConfiguration = pksApiConfiguration.Value ?? throw new ArgumentNullException(nameof(pksApiConfiguration)); ;
```

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common.UnitTests/PermitDecryption/PermitDecryptionTests.cs [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+namespace UKHO.PeriodicOutputService.Common.UnitTests.PermitDecryption
+{
+    [TestFixture]
+    public class PermitDecryptionTests
+    {
+        private ILogger<Common.PermitDecryption.PermitDecryption> fakeLogger;
+        private IOptions<PermitConfiguration> fakePermitConfiguration;
+        private IS63Crypt fakeS63Crypt;
+
+        private IPermitDecryption permitDecryption;
```
Comment (Nadeem14456) [inline UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common.UnitTests/PermitDecryption/PermitDecryptionTests.cs:19]:
```suggestion
        private IPermitDecryption permitDecryption;
        
```

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/IFileSystemHelper.cs [hunk 10]
```diff
@@ -41,5 +41,9 @@ public interface IFileSystemHelper
         bool CreateEmptyFileContent(string filePath);
 
         bool DownloadReadmeFile(string filePath, Stream fileStream);
+
+        Task CreateXmlFromObject<T>(T obj, string filePath, string fileName);
```
Comment (Nadeem14456) [inline UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/IFileSystemHelper.cs:45]: do we need Task here? can we make it void

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/FileSystemHelper.cs [hunk 11]
```diff
@@ -119,6 +122,28 @@ public void CreateXmlFile(byte[] fileContent, string targetPath)
             _fileUtility.CreateXmlFile(fileContent, targetPath);
         }
 
+        public Task CreateXmlFromObject<T>(T obj, string filePath, string fileName)
```
Comment (Nadeem14456) [inline UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/FileSystemHelper.cs:125]: Suggestion: this method can be made generic. just pass the filename without extension.

File: UKHO.PeriodicOutputService/UKHO.BESS.BuilderService/Services/BuilderService.cs [hunk 12]
```diff
@@ -29,13 +32,17 @@ public class BuilderService : IBuilderService
         private readonly IAzureTableStorageHelper azureTableStorageHelper;
         private readonly IOptions<FssApiConfiguration> fssApiConfig;
         private readonly IPksService pksService;
+        private readonly IPermitDecryption permitDecryption;
 
         private const string BESPOKE_FILE_NAME = "V01X01";
         private const string BessBatchFileExtension = "zip";
+        private const string PermitTextFile = "Permit.txt";
+        private const string PermitXmlFile = "Permit.xml";
```
Comment (Nadeem14456) [inline UKHO.PeriodicOutputService/UKHO.BESS.BuilderService/Services/BuilderService.cs:40]: Suggestion: can we decouple extension from filename?
```suggestion
        private const string PERMIT_FILE_NAME= "Permit";
```

File: UKHO.PeriodicOutputService/UKHO.BESS.BuilderService/Program.cs [hunk 13]
```diff
@@ -146,6 +149,12 @@ private static HostBuilder BuildHostConfiguration()
                  serviceCollection.AddScoped<IZipHelper, ZipHelper>();
                  serviceCollection.AddScoped<IFileUtility, FileUtility>();
                  serviceCollection.AddScoped<IAzureTableStorageHelper, AzureTableStorageHelper>();
+                 serviceCollection.AddScoped<IPermitDecryption, PermitDecryption>();
+                 serviceCollection.AddScoped<IS63Crypt, S63Crypt>();
+
+                 serviceCollection.AddTransient<IPksApiClient, PksApiClient>();
+                 serviceCollection.AddSingleton<IAuthPksTokenProvider, AuthTokenProvider>();
+                 serviceCollection.AddScoped<IPksService, PksService>();
```
Comment (Nadeem14456) [inline UKHO.PeriodicOutputService/UKHO.BESS.BuilderService/Program.cs:157]:
```suggestion

```
merge conflict, remove duplicate lines.

**Review by Nadeem14456 — CHANGES_REQUESTED** [review]

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/IFileSystemHelper.cs (hunk 10 above)
Comment (MastekRohitChettri) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/IFileSystemHelper.cs:45]: To achieve unit test added return type as Task.

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/UKHO.PeriodicOutputService.Common.csproj
Comment (MastekRohitChettri) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/UKHO.PeriodicOutputService.Common.csproj:1]: Noted

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Fulfilment.UnitTests/Helpers/FileSystemHelperTests.cs
Comment (MastekRohitChettri) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Fulfilment.UnitTests/Helpers/FileSystemHelperTests.cs:1]: Resolved

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Services/PksService.cs
Comment (MastekRohitChettri) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Services/PksService.cs:1]: Resolved

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/PermitDecryption/S63Crypt.cs (hunk 1 above)
Comment (MastekRohitChettri) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/PermitDecryption/S63Crypt.cs]: Resolved

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/PermitDecryption/BlowfishFactory.cs (hunk 2 above)
Comment (MastekRohitChettri) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/PermitDecryption/BlowfishFactory.cs]: Resolved

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common.UnitTests/PermitDecryption/PermitDecryptionTests.cs (hunk 9 above)
Comment (MastekRohitChettri) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common.UnitTests/PermitDecryption/PermitDecryptionTests.cs:18]: Resolved

File: UKHO.PeriodicOutputService/UKHO.BESS.BuilderService/Program.cs (hunk 13 above)
Comment (MastekRohitChettri) [reply UKHO.PeriodicOutputService/UKHO.BESS.BuilderService/Program.cs]: Resolved

**Review by harshal11869 — CHANGES_REQUESTED** [review]

File: UKHO.FmEssFssMock.API/UKHO.FmEssFssMock.API/Services/FileShareService.cs [hunk 14]
```diff
@@ -33,6 +33,8 @@ public class FileShareService
         private readonly string besMultipleFilesBatchId = "10D40DD5-DDFB-497A-BB67-D99FB1658320";
         private const string BESPOKEREADME = "BESPOKE README";
         private const string MULTIPLEFILES = "MULTIPLE";
+        private const string permitTxtFileName = "Permit.txt";
+        private const string permitXmlFileName = "Permit.xml";
```
[the remark this answers is not in the document]
Comment (MastekRohitChettri) [reply UKHO.FmEssFssMock.API/UKHO.FmEssFssMock.API/Services/FileShareService.cs]: Change to all capital to maintain constant throughout the solution.

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/AuthTokenProvider.cs (hunk 8 above)
Comment (MastekRohitChettri) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/AuthTokenProvider.cs]: Resolved

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/IFileSystemHelper.cs
Comment (MastekRohitChettri) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/IFileSystemHelper.cs:1]: Resolved

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Models/Bess/PKSXml.cs
Comment (MastekRohitChettri) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Models/Bess/PKSXml.cs:1]: Resolved

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Models/Bess/PKSXml.cs (hunk 6 above)
Comment (MastekRohitChettri) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Models/Bess/PKSXml.cs]: Resolved

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Models/Bess/ProductKeyServiceResponse.cs (hunk 5 above)
Comment (MastekRohitChettri) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Models/Bess/ProductKeyServiceResponse.cs]: Resolved

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Models/PermitKey.cs (hunk 4 above)
Comment (MastekRohitChettri) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Models/PermitKey.cs]: Resolved

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/PermitDecryption/BlowfishAlgorithm.cs
Comment (MastekRohitChettri) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/PermitDecryption/BlowfishAlgorithm.cs:1]: Resolved

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/PermitDecryption/BlowfishFactory.cs (hunk 3 above)
Comment (MastekRohitChettri) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/PermitDecryption/BlowfishFactory.cs:8]: Resolved

**Review by Nadeem14456 — CHANGES_REQUESTED** [review]
LGTM.

**Review by Nadeem14456 — APPROVED** [review]

**Review by harshal11869 — APPROVED** [review]
Looks Good.

**Review by Arit14975 — APPROVED** [review]
LGTM

**Review by amogh14893 — APPROVED** [review]
Reviewed DevOps changes. LGTM

**Review by DanOwenUKHO — CHANGES_REQUESTED** [review]
Some comments

File: UKHO.PeriodicOutputService/UKHO.BESS.BuilderService/Services/BuilderService.cs [hunk 15]
```diff
@@ -450,14 +461,57 @@ private async Task<bool> IsBatchCreatedForMock(ConfigQueueMessage configQueueMes
...
             return isBatchCreated;
         }
+
+        private void CreatePermitFile(KeyFileType keyFileType, string filePath, List<ProductKeyServiceResponse> productKeyServiceResponses)
+        {
+            logger.LogInformation(EventIds.PermitFileCreationStarted.ToEventId(), "Permit file creation started for {KeyFileType} | {DateTime} | _X-Correlation-ID : {CorrelationId}", keyFileType, DateTime.UtcNow, CommonHelper.CorrelationID);
+
+            if (keyFileType == KeyFileType.KEY_TEXT)
+            {
+                int i = 1;
```
[the remark this answers is not in the document]
Comment (MastekRohitChettri) [reply UKHO.PeriodicOutputService/UKHO.BESS.BuilderService/Services/BuilderService.cs]: Resolved

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/PermitDecryption/BlowfishAlgorithm.cs [hunk 16]
```diff
@@ -0,0 +1,399 @@
...
+                    _S[j, i + 1] = null1;
+                }
+
+        }
+        //###############################################
+        uint F(uint x)
+        {
+            return (((_S[0, (x >> 24) & 0xFF] + _S[1, (x >> 16) & 0xFF]) ^ _S[2, (x >> 8) & 0xFF]) + _S[3, x & 0xFF]);
+        }
+        //###############################################
```
[the remark this answers is not in the document]
Comment (MastekRohitChettri) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/PermitDecryption/BlowfishAlgorithm.cs]: Resolved

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/PermitDecryption/BlowfishAlgorithm.cs [hunk 17]
```diff
@@ -0,0 +1,399 @@
+﻿using System.Diagnostics.CodeAnalysis;
+
+namespace UKHO.PeriodicOutputService.Common.PermitDecryption
+{
+    [ExcludeFromCodeCoverage]
+    public class BlowfishAlgorithm : IBlowfishAlgorithm
+    {
+        readonly uint[] _P;
```
[the remark this answers is not in the document]
Comment (MastekRohitChettri) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/PermitDecryption/BlowfishAlgorithm.cs:8]: Hi Dan, we just use this file from different repository to decrypt the permit and we did not change anything inside the file. Please revert for any query.

Thanks

File: UKHO.PeriodicOutputService/UKHO.BESS.API.FunctionalTests/Helpers/FssBatchHelper.cs [hunk 18]
```diff
@@ -228,5 +236,94 @@ public static bool CheckInfoFolderAndSerialEncInBessExchangeSet(string? download
...
             }
         }
+
+        /// <summary>
+        /// This method is used to check the Permit files and their content
+        /// </summary>
+        /// <param name="downloadFolderPath">Sets the path of the folder where the required file is downloaded</param>
+        /// <param name="keyFileType">Sets the value KEY_TEXT or PERMIT_XML based on the requested Permit Type</param>
+        /// <returns></returns>
+        public static bool CheckPermitFile(string? downloadFolderPath, string? keyFileType)
```
[the remark this answers is not in the document]
Comment (shivaMastek) [reply UKHO.PeriodicOutputService/UKHO.BESS.API.FunctionalTests/Helpers/FssBatchHelper.cs]: Created two separate methods based on the Permit file being checked.

File: UKHO.PeriodicOutputService/UKHO.BESS.API.FunctionalTests/Helpers/FssBatchHelper.cs [hunk 19]
```diff
@@ -228,5 +236,94 @@ public static bool CheckInfoFolderAndSerialEncInBessExchangeSet(string? download
...
+        {
+            List<string>? cellNames = testConfiguration.bessConfig.ProductsName;
+            List<string>? ck = testConfiguration.bessConfig.Keys;
+            List<string>? cellPermits = testConfiguration.bessConfig.Permits;
+            List<string>? editions = testConfiguration.bessConfig.EditionNumber;
+            string? permitTxt = testConfiguration.bessConfig.PermitTxtFile;
+            string? permitXml = testConfiguration.bessConfig.PermitXmlFile;
+            string? date = testConfiguration.bessConfig.PermitDate;
+
+            if (cellNames != null && ck != null && cellPermits != null && editions != null && permitTxt != null && permitXml != null)
```
[the remark this answers is not in the document]
Comment (shivaMastek) [reply UKHO.PeriodicOutputService/UKHO.BESS.API.FunctionalTests/Helpers/FssBatchHelper.cs]: Scope added.

File: UKHO.PeriodicOutputService/UKHO.BESS.API.FunctionalTests/Helpers/FssBatchHelper.cs [hunk 20]
```diff
@@ -228,5 +236,94 @@ public static bool CheckInfoFolderAndSerialEncInBessExchangeSet(string? download
...
+                                string? encCellEdition = cell.Element("edition")?.Value;
+                                encCellEdition?.Equals(editions[count]).Should().Be(true);
+                                string? encPermit = cell.Element("permit")?.Value;
+                                encPermit?.Equals(cellPermits[count]).Should().Be(true);
+                                count++;
+                            }
+                        }
+                        return true;
+                }
+            return true;
```
[the remark this answers is not in the document]
Comment (shivaMastek) [reply UKHO.PeriodicOutputService/UKHO.BESS.API.FunctionalTests/Helpers/FssBatchHelper.cs]: Modified. Now returning value of variable based in the result of assertions.

File: UKHO.PeriodicOutputService/UKHO.BESS.API.FunctionalTests/Helpers/FssBatchHelper.cs [hunk 21]
```diff
@@ -228,5 +236,94 @@ public static bool CheckInfoFolderAndSerialEncInBessExchangeSet(string? download
...
+
+            var response = await FssApiClient.GetFileDownloadAsync(permitUri);
+
+            if (response.StatusCode == ((HttpStatusCode)200) && keyFileType != null)
+            {
+                Stream permitStream = await response.Content.ReadAsStreamAsync();
+                await using FileStream outputFileStream = new(Path.Combine(batchFolderPath, keyFileType), FileMode.Create);
+                await permitStream.CopyToAsync(outputFileStream);
+            }
+            return true;
```
[the remark this answers is not in the document]
Comment (shivaMastek) [reply UKHO.PeriodicOutputService/UKHO.BESS.API.FunctionalTests/Helpers/FssBatchHelper.cs:336]: Modified. Now returning value of variable based in the result of assertions.

**Pull request merged**